File: problems-adapt/0743_last-to-hear-the-signal/solution_dijkstra.java
import java.util.*;

class Solution {

    public int lastToHear(int[][] edges, int n, int k) {
        Map<Integer, List<int[]>> graph = new HashMap<>();
        for (int[] t : edges) {
            graph
                .computeIfAbsent(t[0], x -> new ArrayList<>())
                .add(new int[] { t[1], t[2] });
        }

        Map<Integer, Integer> dist = new HashMap<>();
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) ->
            Integer.compare(a[0], b[0])
        );
        heap.offer(new int[] { 0, k });
        while (!heap.isEmpty()) {
            int[] top = heap.poll();
            int d = top[0],
                u = top[1];
            // Lazy stale-entry handling: skip nodes settled by an earlier pop.
            if (dist.containsKey(u)) continue;
            // Non-negative weights make the first pop the true shortest
            // distance, so u is final now and never revisited.
            dist.put(u, d);
            List<int[]> outEdges = graph.get(u);
            if (outEdges != null) {
                for (int[] e : outEdges) {
                    if (!dist.containsKey(e[0])) {
                        heap.offer(new int[] { d + e[1], e[0] });
                    }
                }
            }
        }

        // Fewer than n settled nodes means something is unreachable from k.
        if (dist.size() != n) return -1;
        // The last node to hear the signal sets the answer.
        int best = 0;
        for (int v : dist.values()) {
            if (v > best) best = v;
        }
        return best;
    }
}

File: problems-adapt/0201-0300/0245_last-to-hear-the-signal/solution_dijkstra.ts
function lastToHear(edges: number[][], n: number, k: number): number {
    const graph = new Map<number, [number, number][]>(); // node -> [v, w][]
    for (const [u, v, w] of edges) {
        if (!graph.has(u)) graph.set(u, []);
        graph.get(u)!.push([v, w]);
    }

    // min-heap of [dist, node]
    const heap: [number, number][] = [];
    const push = (item: [number, number]): void => {
        heap.push(item);
        let i = heap.length - 1;
        while (i > 0) {
            const p = (i - 1) >> 1;
            if (heap[p][0] <= heap[i][0]) break;
            [heap[p], heap[i]] = [heap[i], heap[p]];
            i = p;
        }
    };
    const pop = (): [number, number] => {
        const top = heap[0];
        const last = heap.pop()!;
        if (heap.length > 0) {
            heap[0] = last;
            let i = 0;
            for (;;) {
                const l = 2 * i + 1,
                    r = 2 * i + 2;
                let m = i;
                if (l < heap.length && heap[l][0] < heap[m][0]) m = l;
                if (r < heap.length && heap[r][0] < heap[m][0]) m = r;
                if (m === i) break;
                [heap[m], heap[i]] = [heap[i], heap[m]];
                i = m;
            }
        }
        return top;
    };

    const dist = new Map<number, number>();
    push([0, k]);
    while (heap.length > 0) {
        const [d, u] = pop();
        // Lazy stale-entry handling: skip nodes settled by an earlier pop.
        if (dist.has(u)) continue;
        // Non-negative weights make the first pop the true shortest distance,
        // so u is final now and never revisited.
        dist.set(u, d);
        const edges = graph.get(u);
        if (edges) {
            for (const [v, w] of edges) {
                if (!dist.has(v)) push([d + w, v]);
            }
        }
    }

    // Fewer than n settled nodes means something is unreachable from k.
    if (dist.size !== n) return -1;
    // The last node to hear the signal sets the answer.
    let best = 0;
    for (const d of dist.values()) {
        if (d > best) best = d;
    }
    return best;
}
